drivers/led/lp3943.c

```diff
@@ -0,0 +1,278 @@
+		break;
+	case 8 ... 11:
+		*reg = LP3943_LS2;
+		*led -= 8;
+		break;
+	case 12 ... 15:
+		*reg = LP3943_LS3;
+		*led -= 12;
+		break;
+	default:
+		SYS_LOG_ERR("Invalid LED specified");
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int lp3943_set_dim_states(struct lp3943_data *data, u32_t led, u8_t mode)
+{
+	int ret;
+	u8_t reg;
+
+	ret = lp3943_get_led_reg(&led, &reg);
+	if (ret) {
+		return ret;
+	}
+
+	/* Set DIMx states for the LEDs */
+	if (i2c_reg_update_byte(data->i2c, CONFIG_LP3943_I2C_ADDRESS, reg,
+				LP3943_MASK << (led << 1),
+				mode << (led << 1))) {
+		SYS_LOG_ERR("LED reg update failed");
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int lp3943_led_blink(struct device *dev, u32_t led,
+			    u32_t delay_on, u32_t delay_off)
+{
+	struct lp3943_data *data = dev->driver_data;
+	struct led_data *dev_data = &data->dev_data;
+	int ret;
+	u16_t period;
+	u8_t reg, val, mode;
+
+	period = delay_on + delay_off;
+
+	if (period < dev_data->min_period || period > dev_data->max_period) {
+		return -EINVAL;
+	}
+
+	/* Use DIM0 for LEDs 0 to 7 and DIM1 for LEDs 8 to 15 */
+	if (led < 8) {
+		mode = LP3943_DIM0;
+	} else {
+		mode = LP3943_DIM1;
+	}
+
+	if (mode == LP3943_DIM0) {
+		reg = LP3943_PSC0;
+	} else {
+		reg = LP3943_PSC1;
+	}
+
+	val = (period * 255) / dev_data->max_period;
+	if (i2c_reg_write_byte(data->i2c, CONFIG_LP3943_I2C_ADDRESS,
+			       reg, val)) {
+		SYS_LOG_ERR("LED write failed");
+		return -EIO;
+	}
+
+	ret = lp3943_set_dim_states(data, led, mode);
+	if (ret) {
+		return ret;
+	}
+
+	return 0;
+}
+
+static int lp3943_led_set_brightness(struct device *dev, u32_t led,
+				     u16_t value)
+{
+	struct lp3943_data *data = dev->driver_data;
+	struct led_data *dev_data = &data->dev_data;
+	int ret;
+	u8_t reg, val, mode;
+
+	if (value < dev_data->min_brightness) {
+		value = dev_data->min_brightness;
+	} else if (value > dev_data->max_brightness) {
+		value = dev_data->max_brightness;
+	}
+
+	/* Use DIM0 for LEDs 0 to 7 and DIM1 for LEDs 8 to 15 */
+	if (led < 8) {
+		mode = LP3943_DIM0;
+	} else {
+		mode = LP3943_DIM1;
+	}
+
+	if (mode == LP3943_DIM0) {
+		reg = LP3943_PWM0;
+	} else {
+		reg = LP3943_PWM1;
+	}
+
+	val = value;
+	if (i2c_reg_write_byte(data->i2c, CONFIG_LP3943_I2C_ADDRESS,
+			       reg, val)) {
+		SYS_LOG_ERR("LED write failed");
+		return -EIO;
+	}
+
+	ret = lp3943_set_dim_states(data, led, mode);
+	if (ret) {
+		return ret;
+	}
+
+	return 0;
+}
+
+static inline int lp3943_led_on(struct device *dev, u32_t led)
+{
+	struct lp3943_data *data = dev->driver_data;
+	int ret;
+	u8_t reg, val, mode;
+
+	ret = lp3943_get_led_reg(&led, &reg);
+	if (ret) {
+		return ret;
+	}
+
+	/* Set LED state to ON */
+	if (i2c_reg_read_byte(data->i2c, CONFIG_LP3943_I2C_ADDRESS,
+			      reg, &val)) {
+		SYS_LOG_ERR("LED read failed");
+		return -EIO;
+	}
+
+	mode = LP3943_ON;
+	val &= ~(LP3943_MASK << (led << 1));
+	val |= (mode << (led << 1));
+	if (i2c_reg_write_byte(data->i2c, CONFIG_LP3943_I2C_ADDRESS,
```

Comment:
Replace this call and the previous i2c call with one call to i2c_reg_update_byte